scripts/hca-utils/hca_utils/utils.py

```diff
@@ -0,0 +1,332 @@
+        :param target_table: The particular table to operate on.
+        :return: A set of row ids to soft delete.
+        """
+        query = f"""
+        SELECT datarepo_row_id
+        FROM `{self.project}.datarepo_{self.dataset}.{target_table}` WHERE file_id IS NULL
+        """
+
+        return self._hit_bigquery(query)
+
+    def _hit_bigquery(self, query):
+        """
+        Helper function to consistently interact with biqquery while reusing the same client.
+        :param query: The SQL query to run.
+        :return: A set of whatever the query is asking for (assumes that we're only asking for a single column).
+        """
+        query_job = self.bigquery_client.query(query)
+        return {row[0] for row in query_job}
+
+    # local csv interactions
+    def create_csv(self, row_ids: Set[str], target_table: str) -> str:
+        """
+        Create a csv locally with one column filled with row ids to soft delete.
+        :param row_ids: A set of row ids to soft delete.
+        :param target_table: The table that the row ids belong to.
+        :return: The filename of the created csv.
+        """
+        filename = self.filename_template.format(table=target_table)
+        with open(filename, mode="w") as soft_delete_file:
+            sd_writer = csv.writer(soft_delete_file, delimiter=",", quotechar='"', quoting=csv.QUOTE_MINIMAL)
+
+            for rid in row_ids:
+                sd_writer.writerow([rid])
+
+        return filename
+
+    @staticmethod
+    def delete_csv(filename: str):
+        """
+        Remove the csv after it has been uploaded so that there isn't a deluge of csv files in the local directory.
+        :param filename: The filename of the file to delete.
+        :return:
+        """
+        os.remove(filename)
+
+    # gcs (cloud storage) interactions
+    def put_csv_in_bucket(self, filename: str) -> str:
+        """
+        Puts a local file into a GCS bucket accessible by Jade.
+        :param filename: The filename of the file to upload.
+        :return: The gs-path of the uploaded file.
+        """
+        storage_client = storage.Client(project=self.bucket_project, credentials=self.gcp_creds)
+        bucket = storage_client.bucket(self.bucket)
+        blob = bucket.blob(filename)
+        blob.upload_from_filename(filename)
+
+        filepath = f"gs://{self.bucket}/{filename}"
+        print(f"Put a soft-delete file here: {filepath}")
+
+        return filepath
+
+    # jade interactions
+    @staticmethod
+    @lru_cache()
+    def _get_endpoint(endpoint: str):
+        """
+        Get relevant information, such as the HTTP method, the path, and parameters, for Jade API endpoints.
+        :param endpoint: The endpoint name to get information for.
+        :return: A dictionary containing the endpoint's method, path, and parameters.
+        """
+        url = "https://jade-terra.datarepo-prod.broadinstitute.org/v2/api-docs"
```

Comment:
❓ Do we need to be pulling the endpoint metadata dynamically like this? I think we're using 2 (`applyDatasetDataDeletion` and `enumerateDatasets`), would it be simpler to just hardcode those for now?

Reply:
Probably simpler but I got carried away here and wanted a generic solution in case we were going to add the polling for retrieveJob (to poll until completion) and if we'd want to extend this to also hit the snapshot creation endpoint. Ideally we'd just use the Jade python client but at the time of writing this, I had not yet been provided with the Python client 🤷🏾 I can hard code if desired

Reply:
🤔 well, this is done and it works, I'm ok keeping it.